catkit2/services/thorlabs_cube_motor_kinesis_sim/thorlabs_cube_motor_kinesis_sim.py

```diff
@@ -0,0 +1,109 @@
+        self.motor_type = 0
+        self.serial_number = str(self.config['serial_number']).encode("UTF-8")
+
+        if self.cube_model == 'KDC101':
+            self.motor_type = 27
+        elif self.cube_model == 'TDC001':
+            self.motor_type = 83
+        else:
+            raise ValueError(f"Motor type {self.motor_type} not supported.")
+
+        self.unit = None
+        self.min_position_config = None
+        self.max_position_config = None
+
+        self.command = None
+        self.current_position = None
+
+    def open(self):
+        """Open connection to the motor."""
+        self.log.debug("Serial number: %s", self.serial_number)
+
+        # Read min, max, unit and model from service configuration.
+        self.min_position_config = self.config['min_position']
+        self.max_position_config = self.config['max_position']
+        self.unit = self.config['unit']
+        self.log.debug("Config parameters: min=%f, max=%f, unit=%s", self.min_position_config, self.max_position_config,
+                       self.unit)
+
+        self.command = self.make_data_stream('command', 'float64', [1], 20)
+        self.current_position = self.make_data_stream('current_position', 'float64', [1], 20)
+
+        nominal_position = self.resolve_position('nominal')
+        self.current_position.submit_data(np.array([nominal_position], dtype='float64'))
+
+        self.make_command('home', self.home)
+
+    def main(self):
+        while not self.should_shut_down:
+            try:
+                # Get an update for the motor position.
+                frame = self.command.get_next_frame(10)
+                # Set the current position if a new command has arrived.
+                self.set_current_position(frame.data[0])
+            except RuntimeError:
+                # Timed out. This is used to periodically check the shutdown flag.
+                continue
+
+    def close(self):
+        pass
+
+    def set_current_position(self, position):
+        """
+        Set the motor's current position.
+
+        The unit is given in real-life units (mm if translation, deg if rotation).
+        """
+        if self.min_position_config <= position <= self.max_position_config:
+            self.testbed.simulator.move_stage(stage_id=self.id,
+                                              old_position=self.get_current_position(),
+                                              new_position=position)
+        else:
+            self.log.warning('Motor not moving since commanded position is outside of configured range.')
+            self.log.warning('Position limits: %f %s <= position <= %f %s.',
+                             self.min_position_config, self.unit, self.max_position_config, self.unit)
+
+    def get_current_position(self):
+        position = self.current_position.get()[0]
+        self.log.debug("Current position: %f %s", position, self.unit)
+        return position
+
+    def resolve_position(self, position):
```

Comment:
Indeed, but it serves a different purpose. Note how this `resolve_position` here only exists in the simulated service and not in the hardware service. This is because the proxy method is not available in the service itself, but we needed something to be able to read the nominal position from the service config. We only need this in the simulated service to be able to put something on the `current_position` data stream when it is created, otherwise you get an error the first time this stream is read in `move_relative` of the proxy method since the data stream is still empty. This is not needed in the hardware service since we are able to ask the device itself for its current position when we start the device and can put that on the data stream instead.

It is indeed not the cleanest way to do it, but I couldn't think of another solution to the empty data stream problem after startup of the service. This solution was adapted from the simulated XPS motors:
https://github.com/spacetelescope/catkit2/blob/6c8900abfd9f96e232a6a9749eaeee2760a5dbbe/catkit2/services/newport_xps_q8_sim/newport_xps_q8_sim.py#L25

Reply:
Ok!